src/core/lib/event_engine/windows/windows_endpoint.cc

```diff
@@ -0,0 +1,333 @@
+      executor_->Run([cb = std::move(on_writable), wsa_error]() mutable {
+        cb(WSAErrorToStatusWithMessage(wsa_error, "WSASend", DEBUG_LOCATION));
+      });
+      return;
+    }
+  }
+  // As all is now setup, we can now ask for the IOCP notification. It may
+  // trigger the callback immediately however, but no matter.
+  handle_write_event_.SetCallback(std::move(on_writable));
+  handle_write_event_.SetSliceBuffer(data);
+  socket_->NotifyOnWrite(&handle_write_event_);
+}
+const EventEngine::ResolvedAddress& WindowsEndpoint::GetPeerAddress() const {
+  return peer_address_;
+}
+const EventEngine::ResolvedAddress& WindowsEndpoint::GetLocalAddress() const {
+  return local_address_;
+}
+
+// ---- Handle{Read|Write}Closure
+
+WindowsEndpoint::BaseEventClosure::BaseEventClosure(WindowsEndpoint* endpoint)
+    : endpoint_(endpoint), cb_(&AbortOnEvent) {}
+
+void WindowsEndpoint::HandleReadClosure::Run() {
+  GRPC_EVENT_ENGINE_TRACE("WindowsEndpoint::%p Handling Read Event", endpoint_);
+  absl::Status status;
+  auto* read_info = endpoint_->socket_->read_info();
+  auto cb_cleanup = absl::MakeCleanup([this, &status]() {
+    auto cb = std::move(cb_);
+    cb_ = &AbortOnEvent;
+    cb(status);
+  });
+  if (read_info->wsa_error() != 0) {
+    status = WSAErrorToStatus(read_info->wsa_error(), DEBUG_LOCATION);
+    buffer_->Clear();
+    return;
+  }
+  if (read_info->bytes_transferred() > 0) {
+    GPR_ASSERT(read_info->bytes_transferred() <= buffer_->Length());
+    if (read_info->bytes_transferred() != buffer_->Length()) {
+      buffer_->RemoveLastNBytes(buffer_->Length() -
+                                read_info->bytes_transferred());
+    }
+    GPR_ASSERT(read_info->bytes_transferred() == buffer_->Length());
+    if (grpc_event_engine_trace.enabled()) {
+      for (int i = 0; i < buffer_->Count(); i++) {
+        auto str = buffer_->RefSlice(i).as_string_view();
+        gpr_log(GPR_INFO, "WindowsEndpoint::%p READ (peer=%s): %.*s", this,
+                endpoint_->peer_address_string_.c_str(), str.length(),
+                str.data());
+      }
+    }
+    return;
+  }
+  // Either the endpoint is shut down or we've seen the end of the stream
+  buffer_->Clear();
+  // TODO(hork): different error message if shut down
+  status = absl::UnavailableError("End of TCP stream");
+}
+
+void WindowsEndpoint::HandleWriteClosure::Run() {
+  GRPC_EVENT_ENGINE_TRACE("WindowsEndpoint::%p Handling Write Event",
+                          endpoint_);
+  auto* write_info = endpoint_->socket_->write_info();
+  auto cb = std::move(cb_);
+  cb_ = &AbortOnEvent;
+  absl::Status status;
+  if (write_info->wsa_error() != 0) {
+    status = WSAErrorToStatusWithMessage(write_info->wsa_error(),
+                                         "Error in WSASend", DEBUG_LOCATION);
+  } else {
+    GPR_ASSERT(write_info->bytes_transferred() == buffer_->Length());
+  }
+  endpoint_->executor_->Run(
+      [cb = std::move(cb), status = std::move(status)]() mutable {
```

Comment:
The read callback is made inline without going through the executor again... is there a reason this one also cannot?